vm_insnhelper.c

```diff
@@ -2525,22 +2525,67 @@ rb_iseq_only_kwparam_p(const rb_iseq_t *iseq)
            ISEQ_BODY(iseq)->param.flags.has_block == FALSE;
 }
 
-static inline void
-vm_caller_setup_arg_splat(rb_control_frame_t *cfp, struct rb_calling_info *calling, VALUE ary)
+#define ALLOW_HEAP_ARGV (-2)
+#define ALLOW_HEAP_ARGV_KEEP_KWSPLAT (-3)
+
+static inline bool
+vm_caller_setup_arg_splat(rb_control_frame_t *cfp, struct rb_calling_info *calling, VALUE ary, int max_args)
 {
     vm_check_canary(GET_EC(), cfp->sp);
+    bool ret = false;
 
     if (!NIL_P(ary)) {
         const VALUE *ptr = RARRAY_CONST_PTR_TRANSIENT(ary);
-        long len = RARRAY_LEN(ary), i;
+        long len = RARRAY_LEN(ary);
+        int argc = calling->argc;
 
-        CHECK_VM_STACK_OVERFLOW(cfp, len);
+        if (UNLIKELY(max_args <= ALLOW_HEAP_ARGV && len + argc > VM_ARGC_STACK_MAX)) {
+            /* Avoid SystemStackError when splatting large arrays by storing arguments in
+             * a temporary array, instead of trying to keeping arguments on the VM stack.
+             */
+            VALUE *argv = cfp->sp - argc;
+            VALUE argv_ary = rb_ary_hidden_new(len + argc + 1);
+            rb_ary_cat(argv_ary, argv, argc);
+            rb_ary_cat(argv_ary, ptr, len);
```

Comment:
Does this mean there is always a full copy of the arguments when nb-of-args > VM_ARGC_STACK_MAX? (for any kind of call)

Reply:
This is only for callers that use `CALLER_SETUP_ARG` with `ALLOW_HEAP_ARGV` or `ALLOW_HEAP_ARGV_KEEP_KWSPLAT`.  Currently, that is these type of calls only:

* cfunc
* ifunc
* non-iseq bmethod
* send(*args) # but not send(:meth, *args)
* :symbol.to_proc.call(*args) # but not :symbol.to_proc.call(recv, *args)